### PR TITLE
Ruby group capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by Lucas-Erkana**
**In this project @Lucash2022 , @codeWithHyder @Kaghenimbale   build the following:**
- Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. _Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created._
- All methods visible in the diagram should be implemented:
    - add_item method (in all classes that have that method)
        - should take an instance of the Item class as an input.
        - should add the input item to the collection of items.
        - should add self as a property of the item object (by using the correct setter from the item object).
    - can_be_archived?() in the Item class
        - should return true if published_date is older than 10 years.
        - otherwise, it should return false.
    - move_to_archive() in the Item class
        - should reuse can_be_archived?() method.
        - should change the archived property to true if the result of the can_be_archived?() method is true.
        - should do nothing if the result of the can_be_archived?() method is false.
    - can_be_archived?() in the Book class
        - should override the method from the parent class.
        - should return true if parent's method returns true OR if cover_state equals to "bad".
        - otherwise, it should return false.
    - can_be_archived?() in the MusicAlbum class
        - should override the method from the parent class.
        - should return true if parent's method returns true AND if on_spotify equals true.
        - otherwise, it should return false.
    - can_be_archived?() in the Movie class
        - should override the method from the parent class.
        - should return true if parent's method returns true OR if silent equals true.
        - otherwise, it should return false.
    - can_be_archived?() in the Game class
        - should override the method from the parent class.
        - should return true if parent's method returns true AND if last_played_at is older than 2 years.
        - otherwise, it should return false.
- Add unit tests for all implemented methods.

#### User interface

- Create a main.rb file that will serve as your console app entry-point.
- Your console app, at the start, should:
    - Present the user with a list of options to perform.
    - Let users choose an option.
    - If needed, ask for parameters for the option.
    - Have a way to quit the app.
- The following options should be available:
    - List all books
    - List all music albums
    - List all movies
    - List of games
    - List all genres (e.g 'Comedy', 'Thriller')
    - List all labels (e.g. 'Gift', 'New')
    - List all authors (e.g. 'Stephen King')
    - List all sources (e.g. 'From a friend', 'Online shop')
    - Add a book
    - Add a music album
    - Add a movie
    - Add a game
 - All data should be preserved by saving collections in .json files.

### Database
- Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

And here is the link to our presentation:
https://youtu.be/W6xHl08JJkk

### CODE REVIEW

**Review by BregornOriginal — CHANGES_REQUESTED** [review]
Hi @Team 💯 💯 💯  

Good job so far!

There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

one-step-closer-one-more

***Excellent work until here! you did an excellent capstone with a lot of features. This is one of the best capstones that you will have done in the program, so let's improve it a bit more to be an excellent showcase while you are looking for a job!*** 🥇 


## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: classes/item.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+require 'date'
+
+class Item
+  attr_reader :id, :genre, :author, :source, :label
```
[the remark this answers is not in the document]
Comment (Lucash2022) [reply classes/item.rb]: Thank you for the review. Removed :source at this [commit](https://github.com/Lucash2022/Catalog-Ruby-Group-Capstone/pull/57/commits/ad2e63f3e41dff753f976b212831efbb8237cec1)

File: schema.sql [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  name VARCHAR(50) NOT NULL,
+  PRIMARY KEY (id)
+);
+
+
+CREATE TABLE MusicAlbum (
+  id SERIAL NOT NULL,
+  on_spotify BOOLEAN NOT NULL,
+  genre_id INT NULL,
+  FOREIGN KEY (genre_id) REFERENCES Genre (id),
```
[the remark this answers is not in the document]
Comment (Lucash2022) [reply schema.sql]: Thank you for the review. It have been resolved at this[ commit](https://github.com/Lucash2022/Catalog-Ruby-Group-Capstone/pull/57/commits/518e6e0c2b426a26750bd2f4eef4cfbcb2212854)

File: schema.sql [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+CREATE TABLE IF NOT EXISTS game (
+  id INT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+  multiplayer text NOT NULL,
+  last_played_at date NOT NULL,
+  item_id integer REFERENCES item(id) ON DELETE CASCADE
+);
+
+
+CREATE TABLE IF NOT EXISTS author (
+  id INT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
```
[the remark this answers is not in the document]
Comment (Lucash2022) [reply schema.sql:12]: Thank you for the review. It have been resolved at this[ commit](https://github.com/Lucash2022/Catalog-Ruby-Group-Capstone/pull/57/commits/518e6e0c2b426a26750bd2f4eef4cfbcb2212854)

**Review by tufoinnkuo10 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @TEAM

Your project is complete! Great job for implementing the changes🥇. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

## Good Points 👍
- Descriptive Pull Request✔️ 
- user-friendly app✔️ 
- well tested✔️ ✔️ ✔️ 
- equal contribution✔️ ✔️ 
- works as expected ✔️ 
- No linter errors✔️ ✔️ 

<br>
Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀

t
 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**